consensus/istanbul/backend/engine.go

```diff
@@ -921,10 +921,69 @@ func (sb *backend) snapshot(chain consensus.ChainReader, number uint64, hash com
 		logger.Trace("Stored voting snapshot to disk", "number", snap.Number, "hash", snap.Hash)
 	}
 
+	sb.regen(chain, headers)
+
 	sb.recents.Add(snap.Hash, snap)
 	return snap, err
 }
 
+// regen commits snapshot data to database
+// regen is triggered if there is any checkpoint block in the `headers`.
+// For each checkpoint block, this function verifies the existence of its snapshot in DB and stores one if missing.
+/*
+ Triggered:
+ |   ^                          ^                          ^                          ^  ...|
+     SI                 SI*(last snapshot)                 SI                         SI
+       			   | header1, .. headerN |
+ Not triggered: (Guaranteed SI* was committed before )
+ |   ^                          ^                          ^                          ^  ...|
+     SI                 SI*(last snapshot)                 SI                         SI
+	                            | header1, .. headerN |
+*/
+func (sb *backend) regen(chain consensus.ChainReader, headers []*types.Header) {
+	if !sb.isRestoringSnapshots.Load() && len(headers) > 1 {
+		var (
+			from        = headers[0].Number.Uint64()
+			to          = headers[len(headers)-1].Number.Uint64()
+			start       = time.Now()
+			commitTried = false
+		)
+
+		// Shortcut: No missing snapshot data to be processed.
+		if to-(to%uint64(params.CheckpointInterval)) < from {
+			return
+		}
+
+		sb.isRestoringSnapshots.Store(true)
+		defer func() {
+			sb.isRestoringSnapshots.Store(false)
+		}()
+		for _, header := range headers {
+			var (
+				hn = header.Number.Uint64()
+				hh = header.Hash()
+			)
+			if params.IsCheckpointInterval(hn) {
+				snap, err := sb.snapshot(chain, hn, hh, nil, false)
+				if err != nil {
+					logger.Warn("[Snapshot] Snapshot restoring failed", "len(headers)", len(headers), "from", from, "to", to, "headerNumber", hn)
+					continue
+				}
+				// Store snapshot data if it was not committed before
+				if loadSnap, _ := sb.db.ReadIstanbulSnapshot(hh); loadSnap == nil {
```

Comment:
Can we check ReadIstanbulSnapshot first? `Snap` doesn't need to be generated if it was committed before.

Reply:
Okay, I'll work on that along with the feedback in the thread above.

Reply:
Updated. PTAL.